File: scrapers/youtube_test.py
import requests
import time
import json
import os
import sys
import logging

# ---------- settings ----------
SCRAPERS_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.normpath(os.path.join(SCRAPERS_DIR, ".."))
CONFIG_FILE = os.path.join(PROJECT_ROOT, "config", "youtube_config.json")
OUT_DIR = os.path.join(PROJECT_ROOT, "data", "youtube")
URL = "https://charts.youtube.com/youtubei/v1/browse?alt=json"
RETRIES = 3
PAUSE = 1.0   # YouTube is one big request per country, so go gentle

try:
    with open(CONFIG_FILE, encoding="utf-8") as f:
        CONFIG = json.load(f)
except FileNotFoundError:
    sys.exit(f"ERROR: config file not found: {CONFIG_FILE}")
except json.JSONDecodeError as e:
    sys.exit(f"ERROR: invalid JSON in {CONFIG_FILE}: {e}")

LOCK_FILE = os.path.join(PROJECT_ROOT, ".youtube.lock")
LOG_FILE  = os.path.join(PROJECT_ROOT, "logs", "youtube.log")
os.makedirs(os.path.join(PROJECT_ROOT, "logs"), exist_ok=True)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    handlers=[logging.FileHandler(LOG_FILE), logging.StreamHandler()],
)
log = logging.getLogger(__name__)


def build_body(country):
    # The country lives in the query string, not the URL.
    query = (
        "flags=MusicCharts__enable_apac_and_shorts_charts_expansion"
        "&perspective=PODCAST_SHOW"
        f"&chart_params_country_code={country}"
        "&chart_params_chart_type=PODCAST_SHOWS_BY_WATCH_TIME"
        "&chart_params_period_type=WEEKLY"
    )
    return {
        "context": CONFIG["context"],
        "browseId": "FEmusic_analytics_charts_home",
        "query": query,
    }


def clean_genre(raw):
    # PODCAST_GENRE_TRUE_CRIME -> True Crime
    if not raw or raw == "PODCAST_GENRE_UNSPECIFIED":
        return None
    words = raw.replace("PODCAST_GENRE_", "").split("_")
    return " ".join(w.capitalize() for w in words)


def find_entries(data):
    # Dig down to podcastShowEntries through YouTube's deep nesting.
    sections = data["contents"]["sectionListRenderer"]["contents"]
    for section in sections:
        block = section.get("musicAnalyticsSectionRenderer")
        if block:
            return block["content"]["podcastShows"][0]["podcastShowEntries"]
    return []


def clean_entry(entry):
    thumbs = entry.get("thumbnail", {}).get("thumbnails", [])
    channel_url = (
        entry.get("channelNavigationEndpoint", {})
        .get("urlEndpoint", {})
        .get("url", "")
    )
    return {
        "rank": entry["chartEntryMetadata"]["currentPosition"],
        "youtube_playlist_id": entry["externalPlaylistId"],
        "name": entry["name"],
        "channel": entry.get("channelName", ""),
        "artwork": thumbs[-1]["url"] if thumbs else "",   # last = largest
        "genre": clean_genre(entry.get("primaryGenre")),  # kept, not filtered on
        "channel_url": channel_url,
    }


def fetch_and_save(country):
    body = build_body(country)
    for attempt in range(1, RETRIES + 1):
        try:
            r = requests.post(URL, json=body, timeout=20)
            r.raise_for_status()
            entries = find_entries(r.json())
            shows = [clean_entry(e) for e in entries]

            chart_data = {
                "platform": "youtube",
                "country": country,
                "slug": "top",
                "fetched_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "count": len(shows),
                "shows": shows,
            }

            folder = os.path.join(OUT_DIR, country)
            os.makedirs(folder, exist_ok=True)
            path = os.path.join(folder, "top.json")
            tmp = path + ".tmp"
            with open(tmp, "w", encoding="utf-8") as out:
                json.dump(chart_data, out, ensure_ascii=False, indent=2)
            os.replace(tmp, path)

            return {"ok": True, "country": country, "count": len(shows)}

        except Exception as err:
            if attempt == RETRIES:
                return {"ok": False, "country": country, "error": str(err)}
            time.sleep(2 * attempt)


def main():
    try:
        import fcntl
        _lf = open(LOCK_FILE, "w")
        fcntl.flock(_lf, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except ImportError:
        pass  # Windows dev environment — skip locking
    except OSError:
        sys.exit("ERROR: another instance is already running")

    countries = CONFIG["countries"]
    log.info(f"Fetching {len(countries)} YouTube charts into {OUT_DIR}/ ...")

    ok = fail = 0
    for country in countries:
        res = fetch_and_save(country)
        if res["ok"]:
            ok += 1
            log.info(f"ok   {res['country']}/top: {res['count']} shows")
        else:
            fail += 1
            log.error(f"FAIL {res['country']}/top: {res['error']}")
        time.sleep(PAUSE)

    log.info(f"Done. {ok} ok, {fail} failed.")
    if fail == len(countries):
        log.error("All countries failed — YouTube API may be broken or credentials expired")
    if fail:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scrapers/youtube_test.php
<?php

$scrapersDir = __DIR__;
$projectRoot = realpath($scrapersDir . '/..');
$configFile = $projectRoot . '/config/youtube_config.json';
$outDir = $projectRoot . '/data/youtube';

define('URL', 'https://charts.youtube.com/youtubei/v1/browse?alt=json');
define('RETRIES', 3);
define('PAUSE_SEC', 1); // 1.0 second pause

// --- Load Config ---
if (!file_exists($configFile)) {
    die("ERROR: config file not found: {$configFile}\n");
}
$configJson = file_get_contents($configFile);
$config = json_decode($configJson, true);
if (json_last_error() !== JSON_ERROR_NONE) {
    die("ERROR: invalid JSON in {$configFile}: " . json_last_error_msg() . "\n");
}
if (!isset($config['countries']) || !isset($config['context'])) {
    die("ERROR: missing key in {$configFile}\n");
}

// --- Single Instance Locking ---
$lockFile = $projectRoot . '/.youtube.lock';
$lockHandle = fopen($lockFile, 'w');
if (!$lockHandle || !flock($lockHandle, LOCK_EX | LOCK_NB)) {
    die("ERROR: another instance is already running\n");
}

// --- Logging Setup ---
$logDir = $projectRoot . '/logs';
if (!is_dir($logDir)) mkdir($logDir, 0777, true);
$logFile = $logDir . '/youtube.log';

function logMsg($level, $message) {
    global $logFile;
    $time = date('Y-m-d H:i:s,000'); 
    $line = "{$time} {$level} {$message}\n";
    echo $line;
    file_put_contents($logFile, $line, FILE_APPEND);
}

// --- URL/Body Builder ---
function build_body($country, $context) {
    $query = "flags=MusicCharts__enable_apac_and_shorts_charts_expansion"
           . "&perspective=PODCAST_SHOW"
           . "&chart_params_country_code={$country}"
           . "&chart_params_chart_type=PODCAST_SHOWS_BY_WATCH_TIME"
           . "&chart_params_period_type=WEEKLY";
           
    return [
        "context"  => $context,
        "browseId" => "FEmusic_analytics_charts_home",
        "query"    => $query,
    ];
}

// --- Data Cleaners ---
function clean_genre($raw) {
    // PODCAST_GENRE_TRUE_CRIME -> True Crime
    if (!$raw || $raw === "PODCAST_GENRE_UNSPECIFIED") {
        return null;
    }
    
    $cleaned = str_replace("PODCAST_GENRE_", "", $raw);
    $words = explode('_', $cleaned);
    
    // Capitalize each word, lowercasing the rest (e.g., TRUE -> True)
    $capitalizedWords = array_map(function($w) {
        return ucfirst(strtolower($w));
    }, $words);
    
    return implode(" ", $capitalizedWords);
}

function find_entries($data) {
    // Dig down to podcastShowEntries through YouTube's deep nesting safely
    $sections = $data['contents']['sectionListRenderer']['contents'] ?? [];
    
    foreach ($sections as $section) {
        $block = $section['musicAnalyticsSectionRenderer'] ?? null;
        if ($block) {
            return $block['content']['podcastShows'][0]['podcastShowEntries'] ?? [];
        }
    }
    return [];
}

function clean_entry($entry) {
    $thumbs = $entry['thumbnail']['thumbnails'] ?? [];
    $lastThumb = end($thumbs); // last = largest

    $channelUrl = $entry['channelNavigationEndpoint']['urlEndpoint']['url'] ?? "";

    return [
        "rank"                => $entry['chartEntryMetadata']['currentPosition'] ?? null,
        "youtube_playlist_id" => $entry['externalPlaylistId'] ?? null,
        "name"                => $entry['name'] ?? null,
        "channel"             => $entry['channelName'] ?? "",
        "artwork"             => $lastThumb ? $lastThumb['url'] : "",
        "genre"               => clean_genre($entry['primaryGenre'] ?? null),
        "channel_url"         => $channelUrl,
    ];
}

// --- Main Execution ---
$countries = $config['countries'];
$context = $config['context'];
if (isset($context['capabilities']) && empty($context['capabilities'])) {
    $context['capabilities'] = new stdClass();
}

logMsg('INFO', "Fetching " . count($countries) . " YouTube charts into {$outDir}/ ...");

$okCount = 0;
$failCount = 0;

foreach ($countries as $country) {
    $body = build_body($country, $context);
    $jsonBody = json_encode($body);
    
    $success = false;
    $errorMsg = "";
    $shows = [];

    for ($attempt = 1; $attempt <= RETRIES; $attempt++) {
        $ch = curl_init(URL);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonBody);
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            'Content-Type: application/json'
        ]);
        // Mimic the Python script so YouTube doesn't block us
        curl_setopt($ch, CURLOPT_USERAGENT, 'python-requests/2.31.0'); 
        curl_setopt($ch, CURLOPT_TIMEOUT, 20);

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $curlError = curl_error($ch);
       

        if ($httpCode >= 200 && $httpCode < 300 && $response) {
            $data = json_decode($response, true);
            $entries = find_entries($data);
            
            if (empty($entries)) {
                $errorMsg = "200 OK, but no entries found in JSON structure.";
            } else {
                foreach ($entries as $e) {
                    $shows[] = clean_entry($e);
                }

                $chartData = [
                    "platform"   => "youtube",
                    "country"    => $country,
                    "slug"       => "top",
                    "fetched_at" => gmdate("Y-m-d\TH:i:s\Z"),
                    "count"      => count($shows),
                    "shows"      => $shows
                ];

                $folder = $outDir . '/' . $country;
                if (!is_dir($folder)) mkdir($folder, 0777, true);
                
                $path = $folder . '/top.json';
                $tmp = $path . '.tmp';
                
                file_put_contents($tmp, json_encode($chartData, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
                rename($tmp, $path); // Atomic replace

                $success = true;
                break; // Break out of the retry loop!
            }
        } else {
            $errorMsg = $curlError ? $curlError : "HTTP Code $httpCode. Response: " . substr($response, 0, 500);
        }

        // Wait before retrying
        if (!$success && $attempt < RETRIES) {
            sleep(2 * $attempt);
        }
    }

    if ($success) {
        $okCount++;
        logMsg('INFO', "ok   {$country}/top: " . count($shows) . " shows");
    } else {
        $failCount++;
        logMsg('ERROR', "FAIL {$country}/top: {$errorMsg}");
    }

    // Be gentle to YouTube
    sleep(PAUSE_SEC);
}

logMsg('INFO', "Done. {$okCount} ok, {$failCount} failed.");

if ($failCount == count($countries)) {
    logMsg('ERROR', "All countries failed — YouTube API may be broken or credentials expired");
}

// Release the instance lock
flock($lockHandle, LOCK_UN);
fclose($lockHandle);

if ($failCount > 0) {
    exit(1);
}